src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -2328,6 +2328,81 @@ def _process_parameter_stmts(self, nodes, parent):
                     # Ensure the interface to this Symbol is static
                     symbol.interface = StaticInterface()
 
+    def _process_interface_block(self, node, symbol_table, visibility_map):
+        '''
+        Processes a Fortran2003.Interface_Block. If the interface is named
+        and consists only of [module] procedure :: <procedure-list> then a
+        GenericInterfaceSymbol is created. Otherwise, a RoutineSymbol of
+        UnknownFortranType is created.
+
+        :param node: the parse tree for the interface block.
+        :type node: :py:class:`fparser.two.Fortran2003.Interface_Block`
+        :param symbol_table: the table to which to add new symbols.
+        :type symbol_table: :py:class:`psyclone.psyir.symbols.SymbolTable`
+        :param visibility_map: information on any explicit symbol visibilities
+            in the current scope.
+        :type visibility_map: dict[
+            str, :py:class:`psyclone.psyir.symbols.Symbol.Visibility`]
+
+        '''
+        # Fortran 2003 standard R1203 says that:
+        #    interface-stmt is INTERFACE [ generic-spec ]
+        #                   or ABSTRACT INTERFACE
+        # where generic-spec is either (R1207) a generic-name or one
+        # of OPERATOR, ASSIGNMENT or dtio-spec.
+        proc_stmts = walk(node, Fortran2003.Procedure_Stmt)
+        if not isinstance(node.children[0].children[0],
+                          Fortran2003.Name):
+            # This interface does not have a name. Therefore we store it as a
+            # RoutineSymbol with an internal name and with the content of the
+            # interface being kept within an UnknownFortranType. As a result
+            # the visibility and interface details of the RoutineSymbol do not
+            # matter.
+            symbol_table.new_symbol(
+                root_name="_psyclone_internal_interface",
+                symbol_type=RoutineSymbol,
+                datatype=UnknownFortranType(str(node).lower()))
+        else:
+            # This interface has a name.
+            name = node.children[0].children[0].string.lower()
+            vis = visibility_map.get(
+                name, symbol_table.default_visibility)
+            # Attempt to work out which routines this interface includes. We
+            # only support those interfaces which use:
+            #     [MODULE] PROCEDURE :: <name-list>
+            # to specify these.
+            rsymbols = []
+            for proc_stmt in proc_stmts:
+                for routine_name in proc_stmt.children[0].children:
+                    rsymbols.append(routine_name.string)
```

Comment:
We are losing "MODULE" but it is entirely optional according to the standard and thus has no semantic implications. I've therefore elected not to worry about it but clearly I should comment that decision.

Reply:
I think I was talking rubbish:
![image](https://github.com/stfc/PSyclone/assets/17409212/05aa26f0-502b-4821-a5f0-ff52d5aaa566)
I'll have to extend what I have to keep track of it.

Reply:
I did have half a thought that we could perhaps look at the `interface` of each `RoutineSymbol` but then, if it's deferred, we won't know whether it should be just `procedure` or `module procedure` so that doesn't help. I think storing it here is the only way forward, as you say.